GLOSSARY.md

```diff
@@ -0,0 +1,40 @@
+# GitOps Glossary
+
+This glossary accompanies the [GitOps Principles](./PRINCIPLES.md), and other supporting documents in this repository.
+
+- ## Continuous
+
+    "Continuous" is intended to match the industry standard term: [reconciliation](#reconciliation) continues to happen, not that it must be instantaneous.
+
+- ## Declarative Description
+
+    A configuration that describes the desired operating state of a system without specifying procedures for how that state will be acheived. This separates configuration (the desired state) from the implementation (commands, API calls, scripts etc.) used to achieve that state.
+
+- ## Desired State
+
+    The aggregate of all configiuration data needed to start, bootstrap, recreate, run, and recover a system. This does not include data stored within a system, eg. database contents.
+
+- ## Drift
+
+    When a system's actual state changes for any reason other than its [desired state](#desired-state)declarations having changed. When this occurs, the system has drifted from its desired state.
```

Comment:
> For example, by the definition given, if I declare I want 10 replicas, there are 8 replicas, and the system starts another, that's drift.

I disagree with this one. The definition clearly states that drift only occurs when changes to the actual system state are made _that aren't based on the desired state changing_. Changes to achieve the desired state are by definition not drift.

Reply:
> Changes to achieve the desired state are by definition not drift.

The change in my example is not a change to the desired state, it is a change to the running state.

Reply:
That quibbling is a bit beside the point though, which I will try to restate: I don't think it's useful to talk about drift, because all you care about is whether you need to take action to move the running state towards the desired state. It doesn't matter what changed.

Reply:
The current definition "... `any reason`... `drifted from its desired state`" appear to exclude the following scenario:

Convergence from right (actual state) to left (desired state) discussed in https://github.com/open-gitops/documents/pull/14#discussion_r676288778

From @lloydchang to @scottrigby:
> desired state can be reconciled; please see https://youtu.be/fZEFWh4xBEw?t=1135 🙂
>
> "Convergence doesn't just mean left to right. It might in fact mean right to left. It might mean there is some kind of automation that's happening in the runtime system that is doing legitimate things and I'll give you a concrete example of later in the talk, that is doing legitimate things that then need to be reflected back into the git repository. Because not only is the git repository; not only is git the interface for operations, it now becomes the source of truth; it is the desired state, so if we're constantly bringing these things into alignment; if something happens in the runtime environment that should be part of the desired state, we need to ensure that is captured as well."

From @scottrigby to @lloydchang:
> @lloydchang I have been a bit hung up on terminology of convergence vs reconciliation. However after talking with @cdavisafc about this more I agree with you. There is more to be said about the details (like write back agents following rules specified in the intentional declarations, not mixing intentional and automated write back info etc), but that can be addressed in a glossary item, or in best practices or another supporting doc

Reply:
@lloydchang yes it's true in that video and elsewhere @cdavisafc is interested in describing a wider idea of "convergence". However I am very sure she did _not_ mean that changes in actual state should take precedence as the source of truth over the versioned desired state declarations.

The scenario she's describing here is when you install specific software running on the system to help update your desired state declarations for you _based on explicit rules set in the desired state_, e.g., pinning versions based on a semver range, or calculating dates based on a pre-defined range. These are interesting and convenient tools for people who manage desired state but want to specify rules for things without always having to update the results of those rules manually.

It's important to note that precaution should be taken whenever allowing _any_ software to automatically write back to your source of truth. Justin Cormack has a really good description of this in his GitOpsDays presentation: [GitOps In The Next Five Years](https://www.youtube.com/watch?v=GNKb0weP7eU&ab_channel=Weaveworks%2CInc.). He suggests some implementation details like ensuring those changes are written to a separate branch in Git, etc. But I think the overall point stands for any setup that uses software to write back to where you store your desired state.

I was thinking we may want to discuss this more and expand on these points in a forthcoming best practices doc, which the group has been planning to work on after the principles and glossary. However we decided not to include more detailed notes like that in the glossary. WDYT?

Reply:
@squaremo we currently refer to "drifting" in the Reconciliation glossary item. Are you thinking we should use a different term than "drift", clarify the difference between drift and the process of moving towards a desired state, or something else?

> For each change in the actual state, who's to say whether it's drift (which suggests moving away from the desired state), or progress

FWIW this is something that I believe is addressed by the (currently omitted) closed loop principle. I'd love your feedback on the notes in the #21 PR description under "Items left out of this PR" > "Existing principle 5": operating in a closed loop. I think this would answer the "who's to say" question. Software agents on whatever kind of system that take feedback into account must be intelligent enough to determine what to do. In k8s that may mean continuing to try to apply the manifests, or even there it may mean attempting to apply N times before pausing and alerting. Right?

Reply:
On making changes back to git: I would think of this as outside GitOps. In some practical systems you might want to have automation that e.g., recognises when there need to be more replicas, and makes that the new recovery point by committing a change to git. There's no need to explicitly include it or to rule it out here.

Reply:
@scottrigby Convergence _(which happens automatically)_ seems too detailed for `BEST-PRACTICES.md`

Reply:
@squaremo I think your points are mostly covered now by 988c758. WDYT? The only difference here is we believe it's important to note that drift occurs everywhere, but GitOps has a unique way of dealing with it. Perhaps we can come up with better terminology, but so far it's been a useful analogy.

@lloydchang let's follow up on where best to write details about automatically writing back to git, and the caution that should be taken in doing so. We already have plans to add details around the specific caution that should be taken when setting up Git, storage buckets, etc, in order to follow GitOps principles beyond just adopting those tools.